### PR TITLE
Towards full generic 3: nuked Gentoype-based VCF import

### PR DESCRIPTION
**Pull request by cseed**
Builds on part 2: https://github.com/hail-is/hail/pull/2048

### CODE REVIEW

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 1]
```diff
@@ -177,8 +178,56 @@ class RowGenotype(r: Row) extends Genotype {
...
+      val adq = queryOpt[IndexedSeq[Int]]("AD")
+      val dpq = queryOpt[java.lang.Integer]("DP")
+      val gqq = queryOpt[java.lang.Integer]("GQ")
+      val plq = queryOpt[IndexedSeq[Int]]("PL")
+      val gpq = queryOpt[IndexedSeq[Double]]("GQ")
+
+      if (plq.isDefined || gpq.isEmpty)
+        (a: Any) =>
+          new RowGenotype(Row(
+            gtq.map(_ (a)).orNull,
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Genotype.scala]: Is this stuff just temporary code while we wait for unsafe / codegen?

Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: Yes.  This is so terrible my eyes are bleeding.  It is also incredibly slow.  As soon as your unsafe stuff goes in I will start integrating it into these towards_...  branches.

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 2]
```diff
@@ -177,8 +178,56 @@ class RowGenotype(r: Row) extends Genotype {
...
+    if (t == TGenotype)
+      (a: Any) => a.asInstanceOf[Genotype]
+    else {
+      def queryOpt[T >: Null](field: String)(implicit hr: HailRep[T]): Option[(Any) => T] = {
+        try {
+          val (ft, fq0) = t.queryTyped(field)
+          if (ft == hr.typ)
+            Some((a: Any) => fq0(a).asInstanceOf[T])
+          else
+            None
```
Comment (jigold) [inline src/main/scala/is/hail/variant/Genotype.scala]: I assume this is guarding for the case where the field exists but someone has defined it as a different type than the one expected by the Genotype class.

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 3]
```diff
@@ -177,8 +178,56 @@ class RowGenotype(r: Row) extends Genotype {
...
+        }
+      }
+
+      val gtq = queryOpt[java.lang.Integer]("GT")(callHr)
+      val adq = queryOpt[IndexedSeq[Int]]("AD")
+      val dpq = queryOpt[java.lang.Integer]("DP")
+      val gqq = queryOpt[java.lang.Integer]("GQ")
+      val plq = queryOpt[IndexedSeq[Int]]("PL")
+      val gpq = queryOpt[IndexedSeq[Double]]("GQ")
```
Comment (jigold) [inline src/main/scala/is/hail/variant/Genotype.scala]: This should be "GP"

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 4]
```diff
@@ -177,8 +178,56 @@ class RowGenotype(r: Row) extends Genotype {
...
+        (a: Any) => {
+          val gp = gpq.map(_ (a)).orNull
+
+          new RowGenotype(Row(
+            gtq.map(_ (a)).orNull,
+            adq.map(_ (a)).orNull,
+            dpq.map(_ (a)).orNull,
+            gqq.map(_ (a)).orNull,
+            if (gp != null)
+              Genotype.weightsToLinear(gp.toArray)
```
Comment (jigold) [inline src/main/scala/is/hail/variant/Genotype.scala]: Why is this necessary? Shouldn't GP already be Array[Double].

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala [hunk 5]
```diff
@@ -295,20 +155,37 @@ case class GenericRecordReader(callFields: Set[String]) extends HtsjdkRecordRead
...
+                         |  caught $e""".stripMargin)
                 }
               }
             }
 
           try {
-            HtsjdkRecordReader.cast(a, f.typ)
+            var r = HtsjdkRecordReader.cast(a, f.typ)
+
+            // handle diploid
```
Comment (jigold) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: is this `handle haploid`?

File: python/hail/context.py [hunk 6]
```diff
@@ -669,16 +665,6 @@ def import_vcf(self, path, force=False, force_bgz=False, header_file=None, min_p
...
 
-        :param bool store_gq: If True, store GQ FORMAT field instead of computing from PL. Only applies if ``generic=False``.
-
-        :param bool pp_as_pl: If True, store PP FORMAT field as PL.  EXPERIMENTAL. Only applies if ``generic=False``.
-
-        :param bool skip_bad_ad: If True, set AD FORMAT field with
-            wrong number of elements to missing, rather than setting
-            the entire genotype to missing. Only applies if ``generic=False``.
-
-        :param bool generic: If True, read the genotype with a generic schema.
```
Comment (jigold) [inline python/hail/context.py]: Don't delete this -- still a parameter.

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala [hunk 7]
```diff
@@ -85,152 +85,6 @@ abstract class HtsjdkRecordReader[T] extends Serializable {
   def genericGenotypes: Boolean
 }
 
-case class GenotypeRecordReader(vcfSettings: VCFSettings) extends HtsjdkRecordReader[Genotype] {
```
Comment (jigold) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Are you intentionally keeping the one subclass `GenericRecordReader` separate rather than adding the one method `readRecord` back to `HtsjdkRecordReader`?

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala [hunk 8]
```diff
@@ -285,7 +139,13 @@ case class GenericRecordReader(callFields: Set[String]) extends HtsjdkRecordRead
             if (f.name == "GT")
               gt
             else {
-              val x = g.getAnyAttribute(f.name)
+              var x = g.getAnyAttribute(f.name)
+              // getAnyAttribute returns empty list for missing AD, PL
+              if (f.name == "AD" && !g.hasAD)
```
Comment (jigold) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: good catch!

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 5 above)
Comment (jigold) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Also, if this is new functionality, add a test -- or is that what testHaploid was supposed to do?

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 2 above)
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: Correct.  Or we could throw an error.  I think the GenotypeRecordReader did neither (e.g. the PP field).

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 3 above)
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: Good catch!

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 4 above)
Comment (cseed) [inline src/main/scala/is/hail/variant/Genotype.scala]: The fifth row element for RowGenotype is PX, which is expected to be an `IndxedSeq[Int]`:

```
  def _unboxedPX: Array[Int] =
    if (r.isNullAt(4))
      null
    else
      r.getSeq(4).toArray
```

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 5 above)
Comment (cseed) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Not new, I had to port this from GentypeRecordReader to get testHaploid to pass.  I think once all of this is done, we'll be able to generalize `Call` easily to support multi-ploidy and phasing.

File: src/test/scala/is/hail/io/ExportVcfSuite.scala [hunk 9]
```diff
@@ -101,16 +101,6 @@ class ExportVcfSuite extends SparkSuite {
     assert(hc.importVCF(out2).same(vds))
   }
 
-  @Test def testPPs() {
-    val vds = hc.importVCF("src/test/resources/sample.PPs.vcf", ppAsPL = true)
-    val out = tmpDir.createTempFile("exportPPs", ".vcf")
-    vds.exportVCF(out, exportPP = true)
```
Comment (jigold) [inline src/test/scala/is/hail/io/ExportVcfSuite.scala]: Since you didn't delete the `export_pp` option from `export_vcf`, I don't think you should delete this test. Or you could remove the option from export_vcf.

File: src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 5 above)
Comment (jigold) [inline src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Change the comment to be `handle haploid`

File: src/test/scala/is/hail/io/ExportVcfSuite.scala (hunk 9 above)
Comment (tpoterba) [inline src/test/scala/is/hail/io/ExportVcfSuite.scala]: deleting the option seems better

Comment (cseed) [inline src/test/scala/is/hail/io/ExportVcfSuite.scala]: Deleted.  Generic exportVCF forthcoming...

**Comment by cseed** [comment]
OK, addressed again (and rebased).
